### PR TITLE
Rework the ChemMaster's output handling

### LINKED ISSUE #9592: ChemMaster is a hot mess
1. Server doesn't validate the client's inputs which means a client could lock the server up spawning infinity pills.
2. It should not be possible for someone to spam spawn pills in the first place and this needs to have some kind of cooldown period.
3. It should just have multiple messages instead of chucking all of the data on one message using an enum key to sort it out.

### LINKED ISSUE #9646: Suggestion: Enforce that ChemMaster does not leave more than X items on it's tile
## Description
Enforce that ChemMaster does not leave more than X items on it's tile. Do not perform pill/bottle creation operation if resulting total item count will go over limit. Right now nothing stops chemists from creating literal hundreds of items, so much it will lag your game heavily if rightclicked (issue #8967) and makes reusing that ChemMaster for sane batches a nightmare.

For X I suggest something like 60 items (half of duffel bag worth of pills, 1500 units of reagent with 25u pills), still allowing somewhat insane large batches.

**Additional context**
Should go nice with my idea #9593 giving tool to manage floor item count. Should mitigate concerns voiced in #9185.

### PR DESCRIPTION
**Pull request by Illiux**
## About the PR 

This pull request reworks the way the ChemMaster 4000 outputs pills and bottles. It now has an internal inventory that can hold either a pill canister or a bottle. It can either fill a pill canister with a specific number of pills of a specific dose or it can fill a bottle with a specific amount of solution from the buffer.

The initial draft of this PR contains only the changes to the ChemMaster itself. Other changes that may be worth including either in this PR or a subsequent one include:
* Adding empty bottles and pill canisters to the ChemDrobe and/or SmartFridge
* Allow empty bottles and pill canisters to be lathed

This PR should:
* Resolve #9592
* Resolve #9646

It might also address #9593 but the implementation here is a bit different.

**Screenshots**

chem_master_new_ui
chem_master_new_ui_bottle

**Changelog**

:cl: illiux
- tweak: The ChemMaster 4000 now outputs directly into a pill canister or bottle

### CODE REVIEW

**Comment by 0x6273** [comment]
I was never a fan of how the ChemMaster could create infinite bottles without glass, so I like this change. I think the UI could be simpler tho. Maybe get rid of the Output Container part of the UI and the create buttons, and make it so when you insert a canister or bottle it fills it to the amount you have set and then ejects it back out?

**Comment by 0x6273** [comment]
Also, inserting a cardboard box with bottles could fill all the bottles that don't already have somthing in it.

File: Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs
Comment (0x6273) [inline Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:184]: Why? The way it was looks cleaner.

Comment (mirrorcult) [reply Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:184]: the old way is convention as well iirc, this should be reverted

File: Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs
Comment (vulppine) [inline Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:268]:
```suggestion
            if (string.IsNullOrEmpty(label)) 
                return;
```

File: Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs
Comment (vulppine) [inline Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:266]: should be replaced with a public API into LabelSystem but uhhh when somebody codes it

File: Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs
Comment (vulppine) [inline Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:193]: probably should replace with a const in the system instead of it being embedded here

File: Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs
Comment (vulppine) [inline Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:273]:
```suggestion
            MetaData(ent).EntityName = val;
```

File: Resources/Prototypes/Entities/Structures/Machines/chem_master.yml
Comment (vulppine) [inline Resources/Prototypes/Entities/Structures/Machines/chem_master.yml:76]: this could just be a tag instead, since the component itself seems to just be exactly that and doesn't seem to have any other use other than signifying that this entity can't fit in the chemmaster (see `Content.Shared/Whitelist/EntityWhitelist.cs` for more information, specifically the Tags field, and `Content.Shared/Tag/TagComponent.cs` for where you would store your tags)

**Review by vulppine — CHANGES_REQUESTED** [review]
definitely has formatting issues, should give that a second pass, especially with one-line ifs (needs to have a line break between the statement and return, as per convention as mentioned earlier)

**Comment by Illiux** [comment]
> Also, inserting a cardboard box with bottles could fill all the bottles that don't already have somthing in it.

I had it working on things inside inserted cardboard boxes in the initial design. I revised it to what you see here [based on a Discord discussion](https://discord.com/channels/310555209753690112/310555209753690112/1018210094418247841) with @mirrorcult.

> I think the UI could be simpler tho. Maybe get rid of the Output Container part of the UI and the create buttons, and make it so when you insert a canister or bottle it fills it to the amount you have set and then ejects it back out?

I wanted to keep as close as possible to the original interaction pattern and UI. I think this might be a good idea, but it's significant enough and seemingly non-essential enough a change that I'd prefer to leave it to a later PR.

**Review by vulppine — APPROVED** [review]
nothing seems terribly wrong with this functionality wise, but i'm wondering if it's possible to make it so that the input/output sections are tabs instead of it being one wide window? that's mostly just a personal comment - otherwise, seems fine (but probably should get another maintainer to review, just in case)

**Comment by Illiux** [comment]
Hmmm, I think I'd have to do a bit more redesign if it were tabbed. One problem would be you'd not be able to see the buffer while using the output. I think this would be confusing and a little annoying, since you get an error from the machine if you try to make a larger batch than you have buffer for. Maybe You'd put the buffer on the bottom (or top) and tab the rest of the interface?

**Comment by vulppine** [comment]
> Hmmm, I think I'd have to do a bit more redesign if it were tabbed. One problem would be you'd not be able to see the buffer while using the output. I think this would be confusing and a little annoying, since you get an error from the machine if you try to make a larger batch than you have buffer for. Maybe You'd put the buffer on the bottom (or top) and tab the rest of the interface?

yeah, something like that: make it so that the buffer persists throughout the rest of the UI, but the input/output sections have their own different tabs you can switch between, that way you're aware of the buffer's contents throughout whichever one you want to work with

**Comment by Visne** [comment]
UI looks really nice, but please add some margins. I also think the name of the canister/beaker/bottle should be on top of the box, not inside of it next to the pills/reagents.

**Comment by Illiux** [comment]
@vulppine thoughts?

chem-master-input-ui
chem-master-output-ui

**Comment by Illiux** [comment]
With margins added to the inner panels:

chem-master-input-ui
chem-master-output-ui

> I also think the name of the canister/beaker/bottle should be on top of the box, not inside of it next to the pills/reagents.

I think I agree. The previous UI already displayed it like this though, and I don't want to expand the scope of this PR to a general ChemMaster UI rework when it's really about changing how it does its outputs.

**Comment by 0x6273** [comment]
Why was `PillAmount` changed from being the amount of pills, to the dosage? The naming is really confusing here, you have both `PillAmount` and `PillNumber` which sounds like the same thing to me.

**Comment by Illiux** [comment]
> Why was `PillAmount` changed from being the amount of pills, to the dosage? The naming is really confusing here, you have both `PillAmount` and `PillNumber` which sounds like the same thing to me.

Because `BottleAmount` referred to what is now called dosage in the case of bottles, and I wanted to make them consistent. You're right though - I'll just rename both to Dosage.

**Review by Just-a-Unity-Dev — APPROVED** [review]

**Comment by 0x6273** [comment]
> > Why was `PillAmount` changed from being the amount of pills, to the dosage? The naming is really confusing here, you have both `PillAmount` and `PillNumber` which sounds like the same thing to me.
> 
> Because `BottleAmount` referred to what is now called dosage in the case of bottles, and I wanted to make them consistent. You're right though - I'll just rename both to Dosage.

Can you rename amount to dose in `ChemMasterCreatePillsMessage` also so the naming is consistent?

**Pull request merged**
